File: SKIBin_old/app/src/main/java/com/h2Invent/skibin/TabSettingsActivity.java
package com.h2Invent.skibin;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.viewpager.widget.ViewPager;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.util.Log;
import android.widget.Toast;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.google.android.material.tabs.TabLayout;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

public class TabSettingsActivity extends AppCompatActivity implements FragmentConnection.OnItemSelectedListener, FragmentSettings.OnItemSelectedListener {
    private ViewPager mViewPager;
    private ViewPageAdapter mViewPagerAdapter;
    private TabLayout mTabLayout;
    public static JSONObject jsonRes;
    private RequestQueue requestQueue;
    private String requestToken;
    private  String idToken;
    private  String urlToken;
    private String emailToken;
    private String ComToken;
    private static int SCAN_ACTIVITY = 1;
    private static int ORG_SCAN_ACTIVITY = 2;
    public static int prechooseTab = -1;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_tab_settings);
        setViewPager();
        TabLayout tabLayout = (TabLayout) findViewById(R.id.tab);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == SCAN_ACTIVITY) {
            if (resultCode == RESULT_OK) {
                String result = data.getStringExtra("result");
                try {

                    ViewPageAdapter adapter = ((ViewPageAdapter) mViewPager.getAdapter());
                    final FragmentConnection fragmentConnection = (FragmentConnection) mViewPagerAdapter.getItem(0);
                    jsonRes = new JSONObject(result);
                    if(jsonRes.getString("type").equals("ORGANISATION")){
                        //this is the stuff for organisation
                        if (fragmentConnection != null) {
                            fragmentConnection.setOrgText(jsonRes.getString("name"),jsonRes.getString("partner"));
                            fragmentConnection.enableOrgSetting(true);
                            fragmentConnection.showUserInfo(false);
                            fragmentConnection.showOrgInfo(true);
                        }
                    }else if (jsonRes.getString("type").equals("USER")){
                        //this is the stuff for a user with email verfication
                        idToken = jsonRes.getString("token");
                        urlToken = jsonRes.getString("url");
                        if (fragmentConnection != null) {
                            fragmentConnection.activateEmail(true);
                            fragmentConnection.enableScanBtn(true);
                            fragmentConnection.enableOrgSetting(false);
                            fragmentConnection.showUserInfo(true);
                            fragmentConnection.showOrgInfo(false);
                        }
                    }

                } catch (JSONException e) {
                    e.printStackTrace();
                }

            }
            if (resultCode == RESULT_CANCELED) {
                //Write your code if there's no result
                Toast.makeText(getApplicationContext(), R.string.settingsFail, Toast.LENGTH_SHORT).show();
            }
        }

    }

    private void setViewPager() {
        mViewPager = (ViewPager) findViewById(R.id.pager);
        mViewPagerAdapter = new ViewPageAdapter(getSupportFragmentManager());
        mViewPager.setAdapter(mViewPagerAdapter);
        mTabLayout = (TabLayout) findViewById(R.id.tab);
        mTabLayout.setupWithViewPager(mViewPager);
    }

    @Override
    public void scanClicked() {
        Intent intent = new Intent(TabSettingsActivity.this, ScanSettings.class);
        startActivityForResult(intent, SCAN_ACTIVITY);
    }

    @Override
    public void userSaveClicked() {
        saveUserData(jsonRes);
    }

    @Override
    public void resetClicked() {
        resetSettings();
    }

    @Override
    public void userEmailConfirmClicked(String token) {
        recieveToken(token, urlToken, idToken);
    }


    @Override
    public void orgSaveClicked() {
        saveData(jsonRes);
    }

    @Override
    public void init() {
        loadData();
    }

    private void saveData(JSONObject response) {
        SharedPreferences sharedPreferences = getSharedPreferences(SaveSettings.SHARED_PREFS, MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.clear();
        try {
            editor.putBoolean(SaveSettings.IS_ORG, true);
            editor.putString(SaveSettings.ORG_Ansprechpartner, response.getString("partner"));
            editor.putString(SaveSettings.ORG_NAME, response.getString("name"));
            editor.putInt(SaveSettings.ORG_ID, response.getInt("id"));
            editor.putString(SaveSettings.ORG_URL, response.getString("url"));
            editor.apply();
        } catch (JSONException e) {
            e.printStackTrace();
        }

        Toast.makeText(this, R.string.settingsSaved, Toast.LENGTH_SHORT).show();

        onBackPressed();
    }

    private void resetSettings() {
        SharedPreferences sharedPreferences = getSharedPreferences(SaveSettings.SHARED_PREFS, MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.clear();
        editor.apply();
        loadData();
        Toast.makeText(this, R.string.settingsDeleted, Toast.LENGTH_SHORT).show();
        onBackPressed();
    }

    private void saveUserData(JSONObject response) {
        final SharedPreferences sharedPreferences = getSharedPreferences(SaveSettings.SHARED_PREFS, MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.clear();
        try {
            JSONObject obj = jsonRes.getJSONObject("user").getJSONObject("info");
            String userNameS = obj.getString("firstName");
            userNameS = userNameS + " ";
            userNameS = userNameS + obj.getString("lastName");
            editor.putString(SaveSettings.USER_NAME, userNameS);
            editor.putBoolean(SaveSettings.IS_USER, true);
            editor.putString(SaveSettings.USER_ORG, obj.getString("organisation"));
            editor.putString(SaveSettings.USER_TOKEN, jsonRes.getString("token"));
            editor.putString(SaveSettings.USER_EMAIL, obj.getString("email"));
            editor.putString(SaveSettings.USER_CHECKIN_URL, obj.getString("urlCheckinKids"));
            editor.putString(SaveSettings.USER_KINDERLISTE_URL, obj.getString("urlKinderListeHeute"));
            editor.apply();

            requestQueue = Volley.newRequestQueue(this);
            StringRequest request = new StringRequest(Request.Method.POST, urlToken,
                    new Response.Listener<String>() {
                        @Override
                        public void onResponse(String response) {
                            try {
                                JSONObject obj = new JSONObject(response);
                                if (!obj.getBoolean("error")) {
                                    Toast.makeText(getApplicationContext(), R.string.settingsSaved, Toast.LENGTH_SHORT).show();
                                    return;
                                }else {
                                    final SharedPreferences sharedPreferences = getSharedPreferences(SaveSettings.SHARED_PREFS, MODE_PRIVATE);
                                    SharedPreferences.Editor editor = sharedPreferences.edit();
                                    editor.clear();
                                    Toast.makeText(getApplicationContext(), R.string.settingsFail, Toast.LENGTH_SHORT).show();
                                }

                            } catch (Throwable t) {

                                Log.e("My App", "Could not parse malformed JSON: \"" + response + "\"");
                            }
                        }
                    },
                    new Response.ErrorListener() {
                        @Override
                        public void onErrorResponse(VolleyError error) {
                            error.printStackTrace();
                        }
                    }) {
                @SuppressLint("HardwareIds")
                @Override
                protected Map<String, String> getParams() {
                    SharedPreferences sharedPreferences = getSharedPreferences(SaveSettings.SHARED_PREFS, MODE_PRIVATE);
                    Map<String, String> params = new HashMap<String, String>();
                    params.put("token", sharedPreferences.getString(SaveSettings.USER_TOKEN,""));
                    return params;
                }
            };
            requestQueue.add(request);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        onBackPressed();
    }

    private void loadData() {
        SharedPreferences sharedPreferences = getSharedPreferences(MainActivity.SHARED_PREFS, MODE_PRIVATE);
        if (sharedPreferences.getBoolean(SaveSettings.IS_ORG, false)) {
            TabLayout tabLayout = (TabLayout) findViewById(R.id.tab);
            TabLayout.Tab tab = tabLayout.getTabAt(0);
            if (tab != null) {
                tab.select();
            }
            final FragmentConnection fragmentConnection = (FragmentConnection) mViewPagerAdapter.getItem(0);
            fragmentConnection.setOrgText(
                    sharedPreferences.getString(SaveSettings.ORG_NAME, ""),
                    sharedPreferences.getString(SaveSettings.ORG_Ansprechpartner, "")
            );
            FragmentConnection.setIsOrg(true);

        }else {
            ViewPageAdapter adapter = ((ViewPageAdapter) mViewPager.getAdapter());
            final FragmentConnection fragmentConnection = (FragmentConnection) mViewPagerAdapter.getItem(0);
            fragmentConnection.setIsOrg(false);
        }
        if (sharedPreferences.getBoolean(SaveSettings.IS_USER, false)) {
            TabLayout tabLayout = (TabLayout) findViewById(R.id.tab);
            TabLayout.Tab tab = tabLayout.getTabAt(0);
            if (tab != null) {
                tab.select();
            }

            ViewPageAdapter adapter = ((ViewPageAdapter) mViewPager.getAdapter());

            final FragmentConnection fragmentConnection = (FragmentConnection) mViewPagerAdapter.getItem(0);
            fragmentConnection.setUserText(
                    sharedPreferences.getString(SaveSettings.USER_ORG, ""),
                    sharedPreferences.getString(SaveSettings.USER_NAME, ""),
                    sharedPreferences.getString(SaveSettings.USER_EMAIL, "")
            );
            fragmentConnection.setUser(true);

        }else {
            ViewPageAdapter adapter = ((ViewPageAdapter) mViewPager.getAdapter());
            final FragmentConnection fragmentConnection = (FragmentConnection) mViewPagerAdapter.getItem(0);
            fragmentConnection.setUser(false);
        }
    }

    private void recieveToken(final String mailToken, final String urlForTokenUser, final String IDTokenUser) {
        requestQueue = Volley.newRequestQueue(this);
        StringRequest request = new StringRequest(Request.Method.POST, urlForTokenUser,
                new Response.Listener<String>() {
                    @Override
                    public void onResponse(String response) {
                        try {
                            JSONObject obj = new JSONObject(response);
                            if (!obj.getBoolean("error")) {
                                jsonRes = obj;
                                urlToken = obj.getString("urlSave");
                                int index = mViewPager.getCurrentItem();
                                ViewPageAdapter adapter = ((ViewPageAdapter) mViewPager.getAdapter());
                                final FragmentConnection fragmentConnection = (FragmentConnection) mViewPagerAdapter.getItem(index);
                                fragmentConnection.activateEmail(false);
                                String userNameS = "";
                                obj = jsonRes.getJSONObject("user").getJSONObject("info");
                                userNameS = obj.getString("firstName");
                                userNameS = userNameS + " ";
                                userNameS = userNameS + obj.getString("lastName");

                                fragmentConnection.setUserText(
                                        userNameS,
                                        obj.getString("organisation"),
                                        obj.getString("email")
                                );
                                fragmentConnection.enableUSerSetting(true);
                                fragmentConnection.enableScanBtn(true);
                                fragmentConnection.setScanText(getApplicationContext().getResources().getString(R.string.noSetting));

                                return;
                            }

                        } catch (Throwable t) {

                            Log.e("My App", "Could not parse malformed JSON: \"" + response + "\"");
                        }
                        new AlertDialog.Builder(TabSettingsActivity.this)
                                .setTitle("Fehler")
                                .setMessage("Sie haben einen Falschen Code eingegeben. Bitte überprüfen Sie den Bestätigungscode in der E-Mail")

                                // Specifying a listener allows you to take an action before dismissing the dialog.
                                // The dialog is automatically dismissed when a dialog button is clicked.

                                // A null listener allows the button to dismiss the dialog and take no further action.
                                .setNegativeButton(R.string.OkText, null)
                                .setIcon(android.R.drawable.ic_dialog_alert)
                                .show();
                    }
                },
                new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        error.printStackTrace();
                    }
                }) {
            @SuppressLint("HardwareIds")
            @Override
            protected Map<String, String> getParams() {

                Map<String, String> params = new HashMap<String, String>();
                params.put("requestToken", IDTokenUser);
                params.put("confirmationToken", mailToken);
                params.put("device", android.os.Build.MANUFACTURER +" "+ android.os.Build.PRODUCT);
                params.put("os",getAndroidVersion());
                params.put("imei", Settings.Secure.getString(getApplicationContext().getContentResolver(),
                        Settings.Secure.ANDROID_ID));
                return params;
            }
        };
        requestQueue.add(request);
    }
    public String getAndroidVersion() {
        String release = Build.VERSION.RELEASE;
        int sdkVersion = Build.VERSION.SDK_INT;
        return "Android SDK: " + sdkVersion + " (" + release +")";
    }
}

File: SKIBin_old/app/src/main/java/com/h2Invent/skibin/TabSettingsActivity.kt
package com.h2Invent.skibin

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.viewpager.widget.ViewPager
import com.android.volley.Request
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.google.android.material.tabs.TabLayout
import org.json.JSONObject

class TabSettingsActivity : AppCompatActivity(),
    FragmentConnection.OnItemSelectedListener,
    FragmentSettings.OnItemSelectedListener {

    private lateinit var viewPager: ViewPager
    private lateinit var viewPagerAdapter: ViewPageAdapter

    private var scannedJson: JSONObject? = null
    private var requestToken: String = ""
    private var confirmationUrl: String = ""
    private var saveUrl: String = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_tab_settings)
        viewPager = findViewById(R.id.pager)
        viewPagerAdapter = ViewPageAdapter(supportFragmentManager)
        viewPager.adapter = viewPagerAdapter
        findViewById<TabLayout>(R.id.tab).setupWithViewPager(viewPager)
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode != REQUEST_CODE_SCAN) return

        if (resultCode == RESULT_OK) {
            val result = data?.getStringExtra(ScanSettingsActivity.EXTRA_RESULT) ?: return
            val json = JSONObject(result)
            scannedJson = json
            when (json.optString("type")) {
                "ORGANISATION" -> handleOrganisationScan(json)
                "USER" -> handleUserScan(json)
            }
        } else {
            Toast.makeText(this, R.string.settingsFail, Toast.LENGTH_SHORT).show()
        }
    }

    override fun scanClicked() {
        startActivityForResult(Intent(this, ScanSettingsActivity::class.java), REQUEST_CODE_SCAN)
    }

    override fun userSaveClicked() {
        saveUserData(scannedJson ?: return)
    }

    override fun orgSaveClicked() {
        saveOrganisationData(scannedJson ?: return)
    }

    override fun resetClicked() {
        getSharedPreferences(SaveSettings.SHARED_PREFS, MODE_PRIVATE).edit().clear().apply()
        Toast.makeText(this, R.string.settingsDeleted, Toast.LENGTH_SHORT).show()
        loadData()
        finish()
    }

    override fun userEmailConfirmClicked(token: String) {
        receiveToken(token)
    }

    override fun init() {
        loadData()
    }

    private fun handleOrganisationScan(json: JSONObject) {
        viewPagerAdapter.connectionFragment.apply {
            setOrgText(json.optString("name"), json.optString("partner"))
            enableOrgSetting(true)
            showOrgInfo(true)
            showUserInfo(false)
            activateEmail(false)
            setScanText(getString(R.string.noSetting))
        }
    }

    private fun handleUserScan(json: JSONObject) {
        requestToken = json.optString("token")
        confirmationUrl = json.optString("url")
        viewPagerAdapter.connectionFragment.apply {
            activateEmail(true)
            enableScanButton(false)
            enableUserSetting(false)
            showUserInfo(false)
            showOrgInfo(false)
        }
    }

    private fun saveOrganisationData(response: JSONObject) {
        getSharedPreferences(SaveSettings.SHARED_PREFS, MODE_PRIVATE).edit().apply {
            clear()
            putBoolean(SaveSettings.IS_ORG, true)
            putString(SaveSettings.ORG_ANSPRECHPARTNER, response.optString("partner"))
            putString(SaveSettings.ORG_NAME, response.optString("name"))
            putInt(SaveSettings.ORG_ID, response.optInt("id"))
            putString(SaveSettings.ORG_URL, response.optString("url"))
            apply()
        }
        Toast.makeText(this, R.string.settingsSaved, Toast.LENGTH_SHORT).show()
        finish()
    }

    private fun saveUserData(response: JSONObject) {
        val info = response.optJSONObject("user")?.optJSONObject("info") ?: return
        val fullName = listOf(info.optString("firstName"), info.optString("lastName"))
            .filter { it.isNotBlank() }
            .joinToString(" ")

        getSharedPreferences(SaveSettings.SHARED_PREFS, MODE_PRIVATE).edit().apply {
            clear()
            putBoolean(SaveSettings.IS_USER, true)
            putString(SaveSettings.USER_NAME, fullName)
            putString(SaveSettings.USER_ORG, info.optString("organisation"))
            putString(SaveSettings.USER_TOKEN, response.optString("token"))
            putString(SaveSettings.USER_EMAIL, info.optString("email"))
            putString(SaveSettings.USER_CHECKIN_URL, info.optString("urlCheckinKids"))
            putString(SaveSettings.USER_KINDERLISTE_URL, info.optString("urlKinderListeHeute"))
            apply()
        }

        if (saveUrl.isBlank()) {
            Toast.makeText(this, R.string.settingsSaved, Toast.LENGTH_SHORT).show()
            finish()
            return
        }

        val request = object : StringRequest(Method.POST, saveUrl,
            {
                Toast.makeText(this, R.string.settingsSaved, Toast.LENGTH_SHORT).show()
                finish()
            },
            { error -> Toast.makeText(this, error.localizedMessage ?: getString(R.string.settingsFail), Toast.LENGTH_LONG).show() }) {
            override fun getParams(): MutableMap<String, String> = mutableMapOf(
                "token" to getSharedPreferences(SaveSettings.SHARED_PREFS, MODE_PRIVATE)
                    .getString(SaveSettings.USER_TOKEN, "")
                    .orEmpty(),
            )
        }
        Volley.newRequestQueue(this).add(request)
    }

    private fun loadData() {
        val sharedPreferences = getSharedPreferences(SaveSettings.SHARED_PREFS, MODE_PRIVATE)
        val fragment = viewPagerAdapter.connectionFragment

        if (sharedPreferences.getBoolean(SaveSettings.IS_ORG, false)) {
            fragment.setOrgText(
                sharedPreferences.getString(SaveSettings.ORG_NAME, "").orEmpty(),
                sharedPreferences.getString(SaveSettings.ORG_ANSPRECHPARTNER, "").orEmpty(),
            )
            fragment.enableOrgSetting(true)
            fragment.showOrgInfo(true)
            fragment.showUserInfo(false)
        } else {
            fragment.showOrgInfo(false)
            fragment.enableOrgSetting(false)
        }

        if (sharedPreferences.getBoolean(SaveSettings.IS_USER, false)) {
            fragment.setUserText(
                sharedPreferences.getString(SaveSettings.USER_NAME, "").orEmpty(),
                sharedPreferences.getString(SaveSettings.USER_ORG, "").orEmpty(),
                sharedPreferences.getString(SaveSettings.USER_EMAIL, "").orEmpty(),
            )
            fragment.enableUserSetting(true)
            fragment.showUserInfo(true)
            fragment.showOrgInfo(false)
            fragment.activateEmail(false)
            fragment.enableScanButton(true)
            fragment.setScanText(getString(R.string.noSetting))
        } else {
            fragment.showUserInfo(false)
            fragment.enableUserSetting(false)
            fragment.activateEmail(false)
            fragment.enableScanButton(true)
            fragment.setScanText(getString(R.string.noSetting))
        }
    }

    private fun receiveToken(mailToken: String) {
        if (confirmationUrl.isBlank() || requestToken.isBlank()) return
        val request = object : StringRequest(Method.POST, confirmationUrl,
            { response ->
                var success = false
                try {
                    val json = JSONObject(response)
                    if (!json.optBoolean("error")) {
                        scannedJson = json
                        saveUrl = json.optString("urlSave")
                        val info = json.optJSONObject("user")?.optJSONObject("info")
                        if (info != null) {
                            val fullName = listOf(info.optString("firstName"), info.optString("lastName"))
                                .filter { it.isNotBlank() }
                                .joinToString(" ")

                            viewPagerAdapter.connectionFragment.apply {
                                activateEmail(false)
                                setUserText(fullName, info.optString("organisation"), info.optString("email"))
                                enableUserSetting(true)
                                enableScanButton(true)
                                setScanText(getString(R.string.noSetting))
                            }
                            success = true
                        }
                    }
                } catch (exception: Exception) {
                    Log.e(LOG_TAG, "Could not parse JSON", exception)
                }
                if (!success) {
                    AlertDialog.Builder(this)
                        .setTitle(R.string.errorTitle)
                        .setMessage(R.string.emailTokenError)
                        .setNegativeButton(R.string.OkText, null)
                        .show()
                }
            },
            { error -> Toast.makeText(this, error.localizedMessage ?: getString(R.string.settingsFail), Toast.LENGTH_LONG).show() }) {
            @SuppressLint("HardwareIds")
            override fun getParams(): MutableMap<String, String> = mutableMapOf(
                "requestToken" to requestToken,
                "confirmationToken" to mailToken,
                "device" to "${Build.MANUFACTURER} ${Build.PRODUCT}",
                "os" to getAndroidVersion(),
                "imei" to Settings.Secure.getString(contentResolver, Settings.Secure.ANDROID_ID),
            )
        }
        Volley.newRequestQueue(this).add(request)
    }

    private fun getAndroidVersion(): String = "Android SDK: ${Build.VERSION.SDK_INT} (${Build.VERSION.RELEASE})"

    companion object {
        private const val REQUEST_CODE_SCAN = 1001
        private const val LOG_TAG = "TabSettingsActivity"
    }
}
